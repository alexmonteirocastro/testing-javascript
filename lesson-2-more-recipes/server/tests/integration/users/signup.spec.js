import supertest from 'supertest';
import app from '../../../index';
import  { User } from '../../../database/models'

describe('The user sign up tests', () => {


    test('should register a new user',  async () => {
        
        //const response = await supertest(app).get('/api/v1/recipes');

        //console.log(response.body)

        // Arrange
        await User.destroy({ where: {} })

        // Get fake user data 
        const fakeUser = {
            name: 'bahdcoder',
            email: 'bahdcoder@gmail.com',
            password: 'password'
        }

        // Action

        // make post request to signup
        const response = await supertest(app).post('/api/v1/users/signup').send(fakeUser)


        // Assertion 

        // 1. response has the user data
        //console.log(response.body.data)
        expect(response.status).toBe(200);
        expect(response.body.data.user.name).toBe(fakeUser.name);
        expect(response.body.data.user.email).toBe(fakeUser.email);
        expect(response.body.data.access_token).toBeTruthy();

        // 2. The database has a user with the credentials we signed up with
        const userFromDatabase = await User.find({ where:{ email: fakeUser.email } });
        //console.log(userFromDatabase)
        expect(userFromDatabase).toBeTruthy();

    });

    test('should return validation error for duplicate email', async () => {
        
        /// arrange

        // prepare fake data
        const fakeUser = {
            name: 'bahdcoder',
            email: 'bahdcoder@gmail.com',
            password: 'password'
        }

        // clean the databse
        await User.destroy({ where: {} })


        // put user to the database. (resgister a user before hand)
        await User.create(fakeUser);

        /// action

        // POST request to register user with duplicate email
        const response = await supertest(app).post('/api/v1/users/signup').send(fakeUser)
        
        /// assertion

        // 1. make sure the response from the server has a 422 status
        expect(response.status).toBe(422);

        // 2. Make sure that errors from server match the scenario
        console.log(response.body);
        expect(response.body.status).toBe('fail');
        expect(response.body.data).toEqual({ errors: [ 'A user with this email already exists.' ] });
        expect(response.body).toMatchSnapshot();

    });

});