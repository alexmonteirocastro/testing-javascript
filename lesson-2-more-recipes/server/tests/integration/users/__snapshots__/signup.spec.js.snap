// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The user sign up tests should return validation error for duplicate email 1`] = `
Object {
  "data": Object {
    "errors": Array [
      "A user with this email already exists.",
    ],
  },
  "status": "fail",
}
`;
